contracts/RewardController.sol

```diff
@@ -0,0 +1,252 @@
+        rewardToken = IERC20Upgradeable(_rewardToken);
+        startBlock = _startRewardingBlock;
+        epochLength = _epochLength;
+        rewardPerEpoch = _rewardPerEpoch;
+        _transferOwnership(_hatGovernance);
+    }
+
+    function setAllocPoint(address _vault, uint256 _allocPoint) external onlyOwner {        
+        updateVault(_vault);
+        uint256 totalAllocPoint = (globalVaultsUpdates.length == 0) ? _allocPoint :
+        globalVaultsUpdates[globalVaultsUpdates.length-1].totalAllocPoint - vaultInfo[_vault].allocPoint + _allocPoint;
+        if (globalVaultsUpdates.length > 0 &&
+            globalVaultsUpdates[globalVaultsUpdates.length-1].blockNumber == block.number) {
+            // already update in this block
+            globalVaultsUpdates[globalVaultsUpdates.length-1].totalAllocPoint = totalAllocPoint;
+        } else {
+            globalVaultsUpdates.push(VaultUpdate({
+                blockNumber: block.number,
+                totalAllocPoint: totalAllocPoint
+            }));
+        }
+
+        vaultInfo[_vault].allocPoint = _allocPoint;
+    }
+
+    /**
+    * @notice Safe HAT transfer function, transfer rewards from the contract only if there are enough
+    * rewards available.
+    * @param _to The address to transfer the reward to
+    * @param _amount The amount of rewards to transfer
+    * @param _vault The vault address
+   */
+    function safeTransferReward(address _to, uint256 _amount, address _vault) internal {
+        if (rewardToken.balanceOf(address(this)) < _amount)
+            revert NotEnoughRewardsToTransferToUser();
+            
+        rewardToken.safeTransfer(_to, _amount);
+
+        emit SafeTransferReward(_to, _vault, _amount, address(rewardToken));
+    }
+
+    /**
+    * @notice Update the vault's rewardPerShare, not more then once per block
+    * @param _vault The vault's address
+    */
+    function updateVault(address _vault) public {
+        VaultInfo storage vault = vaultInfo[_vault];
+        uint256 lastRewardBlock = vault.lastRewardBlock;
+        if (lastRewardBlock == 0) {
+            vaultInfo[_vault].lastRewardBlock = block.number > startBlock ? block.number : startBlock;
+            return;
+        }
+        if (block.number <= lastRewardBlock) {
+            return;
+        }
+
+        vault.lastRewardBlock = block.number;
+
+        uint256 totalShares = IERC20Upgradeable(_vault).totalSupply();
+
+        if (totalShares != 0) {
+            uint256 reward = getVaultReward(_vault, lastRewardBlock);
+            vault.rewardPerShare += (reward * 1e12 / totalShares);
+        }
+
+        if (globalVaultsUpdates.length != 0) {
+            vaultInfo[_vault].lastProcessedTotalAllocPoint = globalVaultsUpdates.length - 1;
+        }
+    }
+
+    /**
+     * @notice Called by owner to set reward multipliers
+     * @param _rewardPerEpoch reward multipliers
+    */
+    function setRewardPerEpoch(uint256[24] memory _rewardPerEpoch) external onlyOwner {
```

Comment:
so if the rewards has already started, this will treat vaults differently depending on whether they calim their rewards before or after setting this value. 

Need to check how this was handled in v1

Reply:
Well it was not settable in V1 so this problem didn't exist. I can't think of a solution that would keep this functionality and not be very complicated.

One thing we can do is allow changing only the epoch that did not yet start.
Another option is to save all the past rewardPerEpochs and times of change and try to calculate based on that, but this is very complex and might require an array that can get too big...
I guess the first option is best but depends on if this functionality will be sufficient

Reply:
One possible solution:
```
if (block.number < startBlock) { // if rewards have nto started yet, update
   rewardPerEpoch = _rewardPerEpoch;
}  else {

    nextEpoch  = (block.number - startBlock) / epochLength + 1;
     // if rewards are ongoing, update the future rewards but keep past and current
     for (nextEpoch, nextEpoch< TOTAL_NUMBER_OF_EPOCHS, nextEpoch++) {
             rewardPerEpoch[nextEpoch] = _rewardPerEpoch[nextEpoch]
      }
  
} 
```

Reply:
ah that is your first suggestion :)